### PR TITLE
Create sign cookies lambda

### PR DESCRIPTION
**Pull request by MancunianSam**
This lambda provides the response when the front end makes a request to upload.tdr.nationalarchives.gov.uk. The request goes:

Route53 -> Cloudfront -> API Gateway -> Lambda

The lambda carries out the following steps.

1. Decode the incoming json from API gateway and retrieve the access token.
2. Validate the token with the Keycloak library. If this has expired or is otherwise invalid, the lambda returns 401.
3. Create the signed cookies using the private key from the environment variables.
4. Create the response with the correct `Set-Cookies` headers set
5. Set the `Access-Control-Allow-Origin` header based on the environment and origin header.
6. Return a 200 response.

### CODE REVIEW

**Review by TomJKing — COMMENTED** [review]
Looks good. Some comments and questions.

Checked the repo setup. Its missing the `tna-digital-archiving-jenkins` user under Manage Access. Not sure that is intended. Also need to add the licence.

File: README.md
Comment (MancunianSam) [reply README.md:12]: I've added a licence in. The `tna-digital-archiving-jenkins` user is part of the `transfer-digital-records` team which does have write access so that should be ok.

File: src/test/scala/uk/gov/nationalarchives/signcookies/LambdaTest.scala
Comment (MancunianSam) [reply src/test/scala/uk/gov/nationalarchives/signcookies/LambdaTest.scala:91]: I think this is right because it stops requests from other origins accessing the url. I tried it by changing my local front end to run on port 9001 instead of 9000 and it won't connect to the /cookies endpoint any more, it gives a CORS error.

It's not the greatest security thing ever because you can still use localhost:9000 but it's better than nothing I guess.

File: src/test/scala/uk/gov/nationalarchives/signcookies/LambdaTest.scala
Comment (MancunianSam) [reply src/test/scala/uk/gov/nationalarchives/signcookies/LambdaTest.scala]: I didn't do them because I couldn't work out how to change the environment variables in a sane way but there probably should be some. I also thought that the Lambda class was getting a bit big so I've split off the `createResponse` and `createCookies` methods into their own class and merged them. 

We can't really test the `createCookies` method as it's mostly calling SDK stuff and the response is encrypted anyway.
I've added tests in for the different combinations of stage, origin and frontend url.

File: src/test/scala/uk/gov/nationalarchives/signcookies/LambdaTest.scala
Comment (suzannehamilton) [reply src/test/scala/uk/gov/nationalarchives/signcookies/LambdaTest.scala]: Yeah, what Sam's described is the correct behaviour. Modern browsers will block JavaScript access to HTTP responses where the allowed origin doesn't match the current site's domain name.

Should there be an assertion about the allowed hosts header in each of these tests to actually check that behaviour? They currently just check the status code.

File: README.md
Comment (MancunianSam) [reply README.md:3]: Nope, copy and paste error.

File: src/main/scala/uk/gov/nationalarchives/signcookies/Lambda.scala
Comment (suzannehamilton) [inline src/main/scala/uk/gov/nationalarchives/signcookies/Lambda.scala]: I pulled these expiry times out of thin air when I did the spike. 🙂 I think they should be shorter, so they're more like the Keycloak tokens. Might be good to run the values past David anyway.

File: src/main/scala/uk/gov/nationalarchives/signcookies/Lambda.scala
Comment (suzannehamilton) [inline src/main/scala/uk/gov/nationalarchives/signcookies/Lambda.scala]: Did you look into making this match the user's IP? It seems like a good security restriction.

File: src/main/scala/uk/gov/nationalarchives/signcookies/Lambda.scala
Comment (MancunianSam) [reply src/main/scala/uk/gov/nationalarchives/signcookies/Lambda.scala]: The way I've done the front end is to call this endpoint when the page loads and then add a line in to call it again inside Tom's refresh token loop. 
So it it's getting refreshed when the token is getting refreshed, we could have them set to 30 minutes, same as the access token and they'll get refreshed together.

I'll put this in Slack anyway.

**Comment by MancunianSam** [comment]
I've added a licence in. The tna-digital-archiving-jenkins user is part of the transfer-digital-records team which does have write access so that should be ok.

File: src/test/scala/uk/gov/nationalarchives/signcookies/LambdaTest.scala
Comment (MancunianSam) [reply src/test/scala/uk/gov/nationalarchives/signcookies/LambdaTest.scala]: I put it in another comment but I've refactored this a bit so the method that returns the allow origin header is tested separately so we can test all combinations now.

File: README.md
Comment (MancunianSam) [reply README.md:1]: I did originally have them in but I replaced them with `project/plugins.sbt` for some reason. I've put them back.

File: src/test/scala/uk/gov/nationalarchives/signcookies/LambdaTest.scala
Comment (MancunianSam) [reply src/test/scala/uk/gov/nationalarchives/signcookies/LambdaTest.scala]: The comment about testing `createCookies` was wrong, we can decode the policy so I've added in a test for the cookies policy.

File: src/main/scala/uk/gov/nationalarchives/signcookies/Lambda.scala
Comment (MancunianSam) [reply src/main/scala/uk/gov/nationalarchives/signcookies/Lambda.scala]: David is happy with the 30 minutes. I've now worked out how to get the policy json from the create cookies method so I've added in a `TimeUtils` trait in a similar way to the consignment API and added a test in for the expiry times.

File: src/main/scala/uk/gov/nationalarchives/signcookies/Lambda.scala
Comment (MancunianSam) [reply src/main/scala/uk/gov/nationalarchives/signcookies/Lambda.scala]: I didn't realise the sourceIp was sent in the request. I've added in the source IP to the case class and then add it to the policy. The test which is checking for the cookie policy now checks for this as well.

File: README.md
Comment (MancunianSam) [reply README.md:24]: That first response is right because the token is expired. The second one looks like missing credentials in IntelliJ. I've updated the README with the expected response.

**Comment by TomJKing** [comment]
@MancunianSam I'm happy with this. @suzannehamilton Are you ok for me to approve this now?

**Comment by suzannehamilton** [comment]
I haven't done a full review, so I'll let Tom hit the approve button, but the cookie restrictions look good to me now. 👍

**Review by TomJKing — APPROVED** [review]

**Pull request merged**
